File: tests/qt_types_standalone/rust/src/qfontmetrics.rs
// SPDX-FileCopyrightText: 2024 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Laurent Montel <laurent.montel@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::QFont;
use cxx_qt_lib::QFontMetrics;

#[cxx::bridge]
mod qfontmetrics_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/qfontmetrics.h");

        type QFontMetrics = cxx_qt_lib::QFontMetrics;
        type QFont = cxx_qt_lib::QFont;
    }

    extern "Rust" {
        fn clone_qfontmetrics(f: &QFontMetrics) -> QFontMetrics;
        fn constructor_qfontmetrics(f: &QFont) -> QFontMetrics;
    }
}

fn clone_qfontmetrics(p: &QFontMetrics) -> QFontMetrics {
    p.clone()
}

fn constructor_qfontmetrics(f: &QFont) -> QFontMetrics {
    QFontMetrics::from(f)
}

File: crates/cxx-qt-lib/src/gui/qfontmetrics.cpp
// clang-format off
// SPDX-FileCopyrightText: 2024 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Laurent Montel <laurent.montel@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

#include "cxx-qt-lib/qfontmetrics.h"

#include "../assertion_utils.h"

// https://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/text/qfontmetrics.h?h=v5.15.6-lts-lgpl#n147
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/text/qfontmetrics.h?h=v6.2.4#n117
assert_alignment_and_size(QFontMetrics,
                          alignof(::std::size_t),
                          sizeof(::std::size_t));

static_assert(!::std::is_trivially_copy_assignable<QFontMetrics>::value);
static_assert(!::std::is_trivially_copy_constructible<QFontMetrics>::value);

static_assert(!::std::is_trivially_destructible<QFontMetrics>::value);
static_assert(QTypeInfo<QFontMetrics>::isRelocatable);
